File: src/content/docs/topics/cs/coding-problems/2d-dynamic-programming/097-interleaving-string.py
def is_interleave(s1, s2, s3):
    pass  # TODO: implement

def _run_tests():
    assert is_interleave('aabcc', 'dbbca', 'aadbbcbcac') == True
    assert is_interleave('aabcc', 'dbbca', 'aadbbbaccc') == False
    assert is_interleave('', '', '') == True
    assert is_interleave('a', '', 'a') == True
    assert is_interleave('', 'b', 'b') == True
    assert is_interleave('a', 'b', 'abc') == False
    # --- large-input timing ---
    import time as _t
    _s1 = 'a' * 100
    _s2 = 'b' * 100
    _t0 = _t.perf_counter()
    is_interleave(_s1, _s2, _s1 + _s2)
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf is_interleave("a"*100, "b"*100, s1+s2): {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/2d-dynamic-programming/097-interleaving-string.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}

final class Solution {
    func isInterleave(_ s1: String, _ s2: String, _ s3: String) -> Bool {
        // TODO: Implement
        fatalError("TODO: Implement")
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:349fbb984ce96e33d2b29e9b9065ee70968dbb5540e5f4b1242a5c6f1b79d4e1
    expectEqual(Solution().isInterleave("aabcc", "dbbca", "aadbbcbcac"), true, "valid-weave")
    expectEqual(Solution().isInterleave("aabcc", "dbbca", "aadbbbaccc"), false, "invalid-weave")
    expectEqual(Solution().isInterleave("", "abc", "abc"), true, "empty-first")
    expectEqual(Solution().isInterleave("a", "b", "a"), false, "length-mismatch")
    expectEqual(Solution().isInterleave("", "", ""), true, "all-empty")
    // EXCLUDED_VECTOR uppercase: ["A","b","Ab"] | Published inputs use lowercase English letters.
    // EXCLUDED_VECTOR digit: ["a","1","a1"] | Published inputs use lowercase English letters.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
